nautobot/core/utils/git.py

```diff
@@ -93,69 +94,140 @@ def checkout(self, branch, commit_hexsha=None):
         """
         Check out the given branch, and optionally the specified commit within that branch.
 
+        Args:
+            branch (str): A branch name, a tag name, or a (possibly abbreviated) commit identifier.
+            commit_hexsha (str): A specific (possibly abbreviated) commit identifier.
+
+        If `commit_hexsha` is specified and `branch` is either a tag or a commit identifier, they must match.
+        If `commit_hexsha` is specified and `branch` is a branch name, it must contain the specified commit.
+
         Returns:
-            (str, bool): commit_hexsha the repo contains now, whether any change occurred
+            (str, bool): commit_hexsha the repo contains now, and whether any change occurred
         """
         # Short-circuit logic - do we already have this commit checked out?
-        if commit_hexsha and commit_hexsha == self.head:
+        if commit_hexsha and self.head.startswith(commit_hexsha):
```

Comment:
Might be a good idea to write a utility function compare possible git hexsha strings:
```py
def compare_git_hexsha(str1, str2):
    if not set(str1).issubset(string.hexdigits) or not set(str1).issubset(string.hexdigits):
        return False
    if str1 == str2:
        return True
    if len(str1) > 6 and str2.startswith(str1):
        return True
    if len(str2) > 6 and str1.startswith(str2):
        return True
    return False
```

Reply:
https://git-scm.com/book/en/v2/Git-Tools-Revision-Selection#_short_sha_1

Reply:
`self.head` should always be the full non-abbreviated commit hash, so I think the current implementation is fine? Not sure what the more complex utility function would gain us here.